RFC-0003-arrow-flight-connector.md

```diff
@@ -0,0 +1,88 @@
+
+## Background
+
+Apache Arrow Flight is a new general-purpose client-server framework that simplifies high performance transport of large datasets over network interfaces. https://arrow.apache.org/blog/2019/10/13/introducing-arrow-flight/
+
+Presto currently supports different datasources through connectors for each type of datasource. The connectors can use JDBC or some other protocol for datasource connection. If we use Flight client to connect to the different datasources we can improve the performance of the datasource connection. Also we would only need to develop one Flight client that can be used across different datasources instead of developing individual connectors for each type of datasource.
+
+## Proposed Implementation
+
+A template for Flight connector will be created as `presto-base-arrow-flight` module. Since each implementation of flight server will be different, this base module can be extended to create a flight connector that supports a particular implementation. The template will support only read operations.
+
+### High level design
+
+#### Proposed design
+
+![Design diagram](arrow-flight-connector/Presto-flight-client.drawio.png)
+
+The Arrow Flight libraries provide a development framework for implementing a service that can send and receive data streams. Flight APIs are used to retrieve the data from the selected table. The steps below need to be followed. 
+
+1. Create a Flight Client using the host, port, ssl information of the Flight service. 
+    1. Maintain a connection pool of Flight clients.
+    2. Reuse a Flight client from the pool that is already connected to the same location as the new connection.
+    3. Evict connections from the pool using the Least Frequently Used strategy.
+2. Use the Flight Client to connect to the Flight service. 
+3. Create a FlightDescriptor using a command
+    1. The connector template will define an abstract class that returns a byte array.
+    1. The concrete implementation of this connector template will define a class that returns a byte array for the `FlightDescriptor.command` method. 
+    2. The concrete implementation will specify the necessary details for the database connection in the `command` byte array.
+4. Use the FlightDescriptor to obtain a FlightInfo object.
+    1. The FlightInfo object will be got from FlightClient by authenticating to the FlightClient using a bearer token.
+    2. The concrete implementation of the connector is responsible for generating or providing the bearer token.
+5. Use the FlightInfo object to obtain the list of endpoints we will be retrieving data from. 
+6. Use each endpoint to obtain a Ticket. 
+7. Use the Ticket to obtain a Stream. 
+8. Use the Stream to read data. 
+    1. Get VectorSchemaRoot from Flight stream
+    2. For each column in the result, build a Block from the FieldVector
+    3. Return a page with all the constructed Blocks.
+    4. Return new pages until all the Flight streams are read.
```

Comment:
For Prestissimo we could directly convert from Arrow to Velox vectors as they are largely compatible. Please can you add some details about that.

Reply:
@ashkrisk Can you address this?

Reply:
I'd already added those back when I originally added the Prestissimo details:
https://github.com/prestodb/rfcs/blob/7fc2ebbfde03716c001972112acf06b6be75a9d3/RFC-0003-arrow-flight-connector.md?plain=1#L81-L86